community/io/src/main/java/org/neo4j/io/pagecache/impl/SingleFilePageSwapper.java

```diff
@@ -308,7 +433,85 @@ public int write( long filePageId, Page page ) throws IOException
             tryReopen( filePageId, e );
             boolean interrupted = Thread.interrupted();
             // Recurse because this is hopefully a very rare occurrence.
-            int bytesWritten = write( filePageId, page );
+            long bytesWritten = write( filePageId, page );
+            if ( interrupted )
+            {
+                Thread.currentThread().interrupt();
+            }
+            return bytesWritten;
+        }
+    }
+
+    @Override
+    public long write( long startFilePageId, Page[] pages, int arrayOffset, int length ) throws IOException
+    {
+        if ( positionLockGetter != null && hasPositionLock )
+        {
+            try
+            {
+                return writePositionedVectoredToFileChannel( startFilePageId, pages, arrayOffset, length );
+            }
+            catch ( IOException ioe )
+            {
+                throw ioe;
+            }
+            catch ( Exception ignore )
+            {
+                // There's a lot of reflection going on in that method. We ignore everything that can go wrong, and
+                // isn't exactly an IOException. Instead, we'll try our fallback code and see what it says.
+            }
+        }
+        return writePositionVectoredFallback( startFilePageId, pages, arrayOffset, length );
+    }
+
+    private long writePositionedVectoredToFileChannel(
+            long startFilePageId, Page[] pages, int arrayOffset, int length ) throws Exception
+    {
+        long fileOffset = pageIdToPosition( startFilePageId );
+        increaseFileSizeTo( fileOffset + (filePageSize * length) );
+        FileChannel channel = unwrappedChannel( startFilePageId );
+        ByteBuffer[] srcs = convertToByteBuffers( pages, arrayOffset, length );
+        return lockPositionWriteVector( startFilePageId, channel, fileOffset, srcs );
+    }
+
+    private ByteBuffer[] convertToByteBuffers( Page[] pages, int arrayOffset, int length ) throws Exception
+    {
+        ByteBuffer[] buffers = new ByteBuffer[length];
+        for ( int i = 0; i < length; i++ )
+        {
+            Page page = pages[arrayOffset + i];
+            buffers[i] = UnsafeUtil.newDirectByteBuffer( page.address(), filePageSize );
+        }
+        return buffers;
+    }
+
+    private FileChannel unwrappedChannel( long startFilePageId )
+    {
+        StoreChannel storeChannel = channel( startFilePageId );
+        return StoreFileChannel.unwrap( storeChannel );
+    }
+
+    private long lockPositionWriteVector(
+            long filePageId, FileChannel channel, long fileOffset, ByteBuffer[] srcs ) throws IOException
+    {
+        try
+        {
+            synchronized ( positionLock( channel ) )
```

Comment:
What about instead of unwrap, retrieving lock each time we modify `StoreFileChannel` abstraction to provide it for us? then we do not need those unwraps and can cache lock for a storeFileChannel lifetime?